### PR TITLE
Multi phase support

### PR DESCRIPTION
**Pull request by rowleya**
Simple changes that support phases rather than time steps.

An example and test is provided in SpiNNakerManchester/SpiNNakerGraphFrontEnd#142

### CODE REVIEW

**Comment by rowleya** [comment]
This now synchronises the start even when the applications don't use the timer as requested.

File: c_common/front_end_common_lib/include/simulation.h [hunk 1]
```diff
@@ -163,4 +171,11 @@ bool simulation_dma_transfer_done_callback_on(uint tag, callback_t callback);
 //! \param[in] tag: the DMA transfer tag to de-register
 void simulation_dma_transfer_done_callback_off(uint tag);
 
+//! \brief set whether the simulation uses the timer.  By default it will
+//!        be assumed that simulations use the timer unless this function is
+//!        called.
+//! \param[in] sim_uses_timer: Whether the simulation uses the timer (1)
+//!                            or not (0)
+void simulation_set_uses_timer(uint sim_uses_timer);
```
[the remark this answers is not in the document]
Comment (rowleya) [reply c_common/front_end_common_lib/include/simulation.h]: Fixed

File: c_common/front_end_common_lib/src/simulation.c [hunk 2]
```diff
@@ -27,6 +27,10 @@
 #include <spin1_api_params.h>
 #include <spin1_api.h>
 
+// Import things from spin1_api that are not explicitly exposed
+extern void spin1_wfi(void);
+extern uint resume_wait(void);
```
[the remark this answers is not in the document]
Comment (rowleya) [reply c_common/front_end_common_lib/src/simulation.c:38]: Technically this is just bringing these in, but I have added docs anyway

File: c_common/front_end_common_lib/src/simulation.c [hunk 3]
```diff
@@ -172,8 +197,21 @@ static void simulation_control_scp_callback(uint mailbox, uint port) {
...
-        spin1_resume(SYNC_WAIT);
+
+        if (stored_start_function != NULL) {
+            spin1_schedule_callback(synchronise_start, 0, 0, 1);
+        }
+        if (uses_timer) {
+            log_info("Resuming");
+            spin1_resume(SYNC_WAIT);
+        } else {
+            if (event.wait) {
```
[the remark this answers is not in the document]
Comment (rowleya) [reply c_common/front_end_common_lib/src/simulation.c:216]: Some documentation has been added

File: spinn_front_end_common/interface/abstract_spinnaker_base.py [hunk 4]
```diff
@@ -1727,7 +1731,12 @@ def _do_data_generation(self, n_machine_time_steps):
 
         inputs["FirstMachineTimeStep"] = self._current_run_timesteps
         inputs["RunTimeMachineTimeSteps"] = n_machine_time_steps
+        inputs["RunTimeSteps"] = n_machine_time_steps
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: Documentation added to explain this.  The clones are there to hide that the same variable is used for steps as for time steps (really time steps are steps that are physically timed, so they are much the same)

File: spinn_front_end_common/interface/abstract_spinnaker_base.py [hunk 5]
```diff
@@ -1727,7 +1731,12 @@ def _do_data_generation(self, n_machine_time_steps):
...
+
+        # This is done twice to make things nicer for things which don't have
+        # time steps
         inputs["DataNTimeSteps"] = self._max_run_time_steps
+        inputs["DataNSteps"] = self._max_run_time_steps
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1741]: It's ok, the jedi will return

File: spinn_front_end_common/interface/abstract_spinnaker_base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: " (really time steps are steps that are physically timed, so they are much the same)" that be its own admission. implies that the correct view is to change all TimeSteps to be just Steps? As then youd have 1 vairable which reflects whats going on in the system. if you keep 2, when only 1 is being used in a given runstate, your opening yaself up to developer confusion.

[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: This is correct, but as the comment says, I didn't want to break existing code when all I had to do was add a variable :).

File: spinn_front_end_common/interface/abstract_spinnaker_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1741]: sure i heard of something called the last jedi. So nah, the mallet stick will ensure it dont return. Plus, previous comment shows how we could kill the clones without jedi intervention

File: spinn_front_end_common/interface/abstract_spinnaker_base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: Note that a developer can use either; this is just to allow them to use steps if they want to.

[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: number 8 of https://dzone.com/articles/best-programming-jokes-amp-quotes

id be more pleased if there was a corresponding issue then to flag that we need to do this. Or you go all bates on this and fix it fully.

[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: I don't know that it needs fixing.  There are 2 interfaces in lots of other places (as you requested), so having another one here is surely not a big deal...

[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: the other 2 interfaces were where there was different requirements (params that made no sense in a given view) this is the same data being labelled as 2 things. where you yaself has said there is a common label we could use for both.

[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: This costs almost nothing to do, and saves implementers having to change their models.  We could break everything for everyone, but that isn't a way to make friends...

[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: (cwl) who needs friends when we have enemies which plan our demise every day (looking at you murphy). but thast why i said a issue. so we can make enemies later. lol

[the remark this answers is not in the document]
Comment (rowleya) [reply spinn_front_end_common/interface/abstract_spinnaker_base.py:1736]: OK, if it makes this mergable, so be it.  See #601.

**Pull request merged**